File: 06-sobrecarga-igualdad/python/demo06.py
"""
DEMO 06 - "SOBRECARGA", __str__ Y __eq__
Ejecutar:  python3 demo06.py

ESTO ES LO QUE MAS CONFUNDE AL CAMBIAR DE LENGUAJE:
    Java:    ==  compara referencias  |  equals() compara contenido
    Python:  is  compara referencias  |  ==       compara contenido

Dos archivos:
    fraccion.py   la clase
    demo06.py     el programa   <- estas aqui
"""

from fraccion import Fraccion


if __name__ == "__main__":

    a = Fraccion(1, 2)
    b = Fraccion(3, 4)

    print("--- __str__: como se ve el objeto al imprimirlo ---")
    print(f"a = {a}")
    print(f"b = {b}")
    print("(sin __str__ saldria algo como <fraccion.Fraccion object at 0x7f...>)")

    print("--- 'Sobrecarga': un metodo que revisa el tipo ---")
    print(f"a.sumar(b)  = {a.sumar(b)}")
    print(f"a.sumar(2)  = {a.sumar(2)}")
    print("Python elige con isinstance(), en tiempo de EJECUCION.")

    print("--- is vs == ---")
    x = Fraccion(1, 2)
    y = Fraccion(1, 2)

    print(f"x is y        -> {x is y}")
    print("   (compara si son EL MISMO objeto: el '==' de Java)")
    print(f"x == y        -> {x == y}")
    print("   (compara el CONTENIDO: el 'equals()' de Java)")

    z = x
    print(f"x is z        -> {x is z}   (misma referencia)")

    print("--- __eq__ con fracciones equivalentes ---")
    print(f"1/2 == 2/4    -> {Fraccion(1, 2) == Fraccion(2, 4)}")

    print("--- LO QUE JAVA NO PUEDE HACER: el operador + ---")
    print(f"a + b = {a + b}")
    print(f"a + 2 = {a + 2}")
    print("En Java tendrias que escribir a.sumar(b): no se puede redefinir '+'.")

    print("--- __repr__ vs __str__ ---")
    print(f"str(a)  -> {str(a)}      (para el usuario)")
    print(f"repr(a) -> {repr(a)}   (para el programador)")
    print(f"Dentro de una lista Python usa __repr__: {[a, b]}")

File: 06-sobrecarga-igualdad/java/Demo06.java
/*
 * DEMO 06 - SOBRECARGA, toString() Y equals()
 *
 * Compilar y ejecutar:
 *     javac *.java
 *     java Demo06
 *
 * Dos archivos:
 *     Fraccion.java   la clase
 *     Demo06.java     el programa   <- estas aqui
 */
public class Demo06 {
    public static void main(String[] args) {

        Fraccion a = new Fraccion(1, 2);
        Fraccion b = new Fraccion(3, 4);

        System.out.println("--- toString(): como se ve el objeto al imprimirlo ---");
        System.out.println("a = " + a);          // Java llama a toString() solito
        System.out.println("b = " + b);
        System.out.println("(sin toString saldria algo como Fraccion@1b6d3586)");

        System.out.println("--- SOBRECARGA: mismo nombre, distintos parametros ---");
        System.out.println("a.sumar(b)  = " + a.sumar(b));    // version con Fraccion
        System.out.println("a.sumar(2)  = " + a.sumar(2));    // version con int
        System.out.println("Java elige el metodo segun el TIPO del argumento.");

        System.out.println("--- == vs equals() ---");
        Fraccion x = new Fraccion(1, 2);
        Fraccion y = new Fraccion(1, 2);

        System.out.println("x == y        -> " + (x == y));
        System.out.println("   (compara si son EL MISMO objeto en memoria: no lo son)");
        System.out.println("x.equals(y)   -> " + x.equals(y));
        System.out.println("   (compara el CONTENIDO, porque escribimos equals())");

        Fraccion z = x;
        System.out.println("x == z        -> " + (x == z) + "   (misma referencia)");

        System.out.println("--- equals() con fracciones equivalentes ---");
        Fraccion mitad = new Fraccion(1, 2);
        Fraccion dosCuartos = new Fraccion(2, 4);
        System.out.println("1/2 equals 2/4 -> " + mitad.equals(dosCuartos)
                           + "   (comparamos en cruz: 1*4 == 2*2)");
    }
}
